src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,1052 @@
+
+    for (size_t j = 1; j < InImage.n_rows-1; j++)
+      dy.slice(i).row(j) = 0.5 * ( InImage.slice(i).row(j+1) - InImage.slice(i).row(j) );
+  }  
+
+  CubeType mag(InImage.n_rows, InImage.n_cols, InImage.n_slices);
+  for (size_t i = 0; i < InImage.n_slices; ++i)
+  {
+    mag.slice(i) = arma::sqrt( arma::square \
+                  ( dx.slice(i) + arma::square( dy.slice(i) ) ) );
+  }
+
+  arma::umat Location(InImage.n_rows, InImage.n_cols);
+  this->MaxAndLoc(mag, Location, Magnitude);
+  if(grdNormRad != 0)
+  {
+    //we have to do this or override ConvTriangle and SepFilter2D methods.
+    CubeType mag2(InImage.n_rows, InImage.n_cols, 1);
+    mag2.slice(0) = Magnitude;
+    this->ConvTriangle(mag2, grdNormRad);
+    Magnitude = Magnitude / (mag2.slice(0) + 0.01);
+  }
+  MatType dx_mat(dx.n_rows, dx.n_cols),\
+            dy_mat(dy.n_rows, dy.n_cols);
+
+  for(size_t j = 0; j < InImage.n_cols; ++j)
+  {
+    for(size_t i = 0; i < InImage.n_rows; ++i)
+    {
+      dx_mat(i, j) = dx(i, j, Location(i, j));
+      dy_mat(i, j) = dy(i, j, Location(i, j));
+    }
+  }
+  Orientation = arma::atan(dy_mat / dx_mat);
+
+  Orientation.transform( [](double val)\
+   { if(val < 0) return (val + arma::datum::pi);\
+     else return (val);} );
+  
+  for(size_t j = 0; j < InImage.n_cols; ++j)
+  {
+    for(size_t i = 0; i < InImage.n_rows; ++i)
+    {
+      if( abs(dx_mat(i, j)) + abs(dy_mat(i, j)) < 1E-5)
+        Orientation(i, j) = 0.5 * arma::datum::pi;
+    }
+  }
+}
+
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+Histogram(const MatType& Magnitude,
+          const MatType& Orientation, 
+          size_t downscale, size_t interp,
+          CubeType& HistArr)
+{
+
+  //numOrient: number of orientations per gradient scale
+  const size_t numOrient = this->params.NumOrient();
+  const size_t n_rbin = (Magnitude.n_rows + downscale - 1) / downscale;
+  const size_t n_cbin = (Magnitude.n_cols + downscale - 1) / downscale;
+
+  double o_range, o;
+  o_range = arma::datum::pi / numOrient;
+
+  HistArr = CubeType(n_rbin, n_cbin, numOrient);
+  HistArr.zeros();
+
+  size_t r, c, o1, o2;
+  for(size_t i = 0; i < Magnitude.n_rows; ++i)
+  {
+    for(size_t j = 0; j < Magnitude.n_cols; ++j)
+    {
+      r = i / downscale;
+      c = j / downscale;
+
+      if( interp != 0)
+      {
+        o = Orientation(i, j) / o_range;
+        o1 = ((size_t) o) % numOrient;
+        o2 = (o1 + 1) % numOrient;
+        HistArr(r, c, o1) += Magnitude(i, j) * (1 + (int)o - o);
+        HistArr(r, c, o2) += Magnitude(i, j) * (o - (int) o);
+      }
+      else
+      {
+        o1 = (size_t) (Orientation(i, j) / o_range + 0.5) % numOrient;
+        HistArr(r, c, o1) += Magnitude(i, j);
+      }
+    }
+  }
+
+  HistArr = HistArr / downscale;
+
+  for (size_t i = 0; i < HistArr.n_slices; ++i)
+    HistArr.slice(i) = arma::square(HistArr.slice(i));
+}
```

Comment:
arma::square works for Cube<T> too, following should work

`HistArr = arma::square(HistArr);`